reason-parser/src/reason_parser.mly

```diff
@@ -1757,80 +1822,83 @@ signature:
 
 signature_item: mark_position_sig(_signature_item) { $1 }
 _signature_item:
-    LET as_loc(val_ident) COLON core_type post_item_attributes {
+    post_item_attributes LET as_loc(val_ident) COLON core_type {
         let loc = mklocation $symbolstartpos $endpos in
-        let core_type_loc = mklocation $startpos($4) $endpos($4) in
-        mksig(Psig_value (Val.mk $2 (only_core_type $4 core_type_loc) ~attrs:$5 ~loc))
+        let core_type_loc = mklocation $startpos($5) $endpos($5) in
+        mksig(Psig_value (Val.mk $3 (only_core_type $5 core_type_loc) ~attrs:$1 ~loc)) ~loc
     }
-  | EXTERNAL as_loc(val_ident) COLON core_type EQUAL primitive_declaration post_item_attributes {
+  | post_item_attributes EXTERNAL as_loc(val_ident) COLON core_type EQUAL primitive_declaration {
       let loc = mklocation $symbolstartpos $endpos in
-      let core_type_loc = mklocation $startpos($4) $endpos($4) in
-      mksig(Psig_value (Val.mk $2 (only_core_type $4 core_type_loc) ~prim:$6 ~attrs:$7 ~loc))
+      let core_type_loc = mklocation $startpos($5) $endpos($5) in
+      mksig(Psig_value (Val.mk $3 (only_core_type $5 core_type_loc) ~prim:$7 ~attrs:$1 ~loc)) ~loc
     }
   | many_type_declarations {
-        let (nonrec_flag, tyl) = $1 in
-        mksig(Psig_type (nonrec_flag, List.rev tyl))
+      let (nonrec_flag, tyl) = $1 in
+      let hd = List.hd (List.rev tyl) in
+      mksig(Psig_type (nonrec_flag, List.rev tyl)) ~loc:hd.ptype_loc
```

Comment:
I had quite some challenges with the handling of whitespace and this resulted in some not so pretty hacks. This is definitely one of them, however I'm not sure how to improve upon it without breaking the handling of whitespace.